### DISCUSSION TITLE
Best way to store aggregated data

### DISCUSSION THREAD

**Discussion by rottensunday** [ClickHouse/ClickHouse#53846] [Q&A]
Hello everyone. I'm having trouble designing my views so maybe you can help me.
Basically I have a big table with events (billions of entries). These events are triggered by users. Some events might be transactions.
Now I want to quickly iterate over users and find some info about them: number of events, earliest event, number of events in last 30 days, 7 days, median value of transactions in last 7 days, I also want to make some joins to current prices. I want to also filter out users based on some of these values.
Logical solution would be to use Materialized View using AggregatingMergeTree engine. But I have some issues with that:
1. If there are no new events for given user, his data like 'number of events in last 30 days' would be incorrect, as materialized view is an insert trigger
2. there will be an issue with JOINs for same reason as in previous point. If prices change and user won't trigger any events, his 'current value of items' data will be incorrect

Second option I'm thinking about is to use engine ReplacingMergeTree into which I would insert aggregated reports about users every hour or even less often. From what I understand ReplacingMergeTree is supposed to replace entries based on key, so I would always have newest one. And I won't have issues stated previously as I would have full control over how and when I do aggregations. But I don't know if this is suitable solution and it will work correctly?

Any advice would be appreciated.

**Comment by den-crane** [comment]
The first is preferred. You need to add some `date` dimension into your MV, like `hour` or `day` then you will be able to query `last 30 days`.
